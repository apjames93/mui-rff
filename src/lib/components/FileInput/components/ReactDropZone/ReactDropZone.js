import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import Dropzone from 'react-dropzone';

export const base64File = (file, field) => {
    const reader = new FileReader(); // eslint-disable-line
  return new Promise((resolve) => {
    reader.onload = (event) => {
      field.input.onChange(event.target.result);
      resolve(event.target.result);
    };
    reader.readAsDataURL(file);
  });
};

export const drop = async (acceptedFiles, rejectedFiles, field) => {
  if (rejectedFiles.length > 0) {
    const error = rejectedFiles[0];
    field.input.onChange(error);
  }

  if (acceptedFiles.length === 1) {
    await base64File(acceptedFiles[0], field);
  }

  if (acceptedFiles.length > 1) {
    field.input.onChange({
      errors: [{
        code: 'one file',
        message: 'only one file can be uploaded',
      }],
    });
  }
};

const ReactDropZone = ({
  acceptFileTypes,
  onDrop,
  field,
}) => (
  <Dropzone
    onDrop={(acceptedFiles, rejectedFiles) => onDrop(acceptedFiles, rejectedFiles, field)}
    accept={acceptFileTypes.join(', ')}
    data-testid="DropzoneInput"
  >
    {({ getRootProps, getInputProps }) => (
      <div {...getRootProps()}>
        <input {...getInputProps()} />
        <Typography variant="subtitle1">
          Drag &#39;n&#39; drop some files here, or click to select files
        </Typography>
      </div>
    )}
  </Dropzone>
);
export default ReactDropZone;

ReactDropZone.propTypes = {
  acceptFileTypes: PropTypes.arrayOf(PropTypes.string),
  onDrop: PropTypes.func,
  field: PropTypes.shape({
    input: PropTypes.shape({
      onChange: PropTypes.func.isRequired,
    }),
  }).isRequired,
};

ReactDropZone.defaultProps = {
  acceptFileTypes: ['image/jpeg', 'image/png', 'application/pdf'],
  onDrop: drop,
};
