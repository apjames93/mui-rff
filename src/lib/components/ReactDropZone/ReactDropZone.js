import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import Dropzone from 'react-dropzone';
import Thumbs from './components/Thumbs/Thumbs';

const ReactDropZone = ({
  acceptFileTypes,
  onDrop,
  field,
}) => {
  const [state, setState] = useState({ files: {} });

  return (
    <Dropzone
      onDrop={
        (acceptedFiles, rejectedFiles) => onDrop(acceptedFiles, rejectedFiles, field, setState)
      }
      accept={acceptFileTypes.join(', ')}
      data-testid="DropzoneInput"
    >
      {({ getRootProps, getInputProps }) => (
        <div {...getRootProps()}>
          <input {...getInputProps()} />
          <Typography variant="subtitle1">
            Drag &#39;n&#39; drop some files here, or click to select files
          </Typography>
          <Thumbs files={state.files} />
        </div>
      )}
    </Dropzone>
  );
};
export default ReactDropZone;

ReactDropZone.propTypes = {
  acceptFileTypes: PropTypes.arrayOf(PropTypes.string),
  /**
   * callBack function to set Thumbs to files state and bind data to rff
   */
  onDrop: PropTypes.func.isRequired,
  /**
   * rff field
   */
  field: PropTypes.shape({
    input: PropTypes.shape({
      onChange: PropTypes.func.isRequired,
    }),
  }).isRequired,
};

ReactDropZone.defaultProps = {
  acceptFileTypes: ['image/jpeg', 'image/png', 'application/pdf'],
};
