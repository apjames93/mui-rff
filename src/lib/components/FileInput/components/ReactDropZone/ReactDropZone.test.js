import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import ReactDropZone, { drop } from './ReactDropZone';

function setup() {
  const props = {
    onDrop: jest.fn(),
    field: {
      input: {
        onChange: jest.fn(),
      },
    },
  };
  const comp = mount(<ReactDropZone {...props} />);
  return { comp, props };
}


test('snapshot', () => {
  const tree = renderer.create(
    <ReactDropZone />,
  );
  expect(tree).toMatchSnapshot();
});

test('onDrop', () => {
  const { comp, props } = setup();
  comp.find('[data-testid="DropzoneInput"]').first().props().onDrop();
  expect(props.onDrop).toBeCalled();
});

describe('drop', () => {
  test('drop rejectedFiles', async () => {
    const { props } = setup();
    const rejectedFiles = [{
      code: 'this is bad',
      message: 'real bad',
    }];
    await drop([], rejectedFiles, props.field);
    expect(props.field.input.onChange).toBeCalledWith(rejectedFiles[0]);
  });

  test('acceptedFiles.length > 1', async () => {
    const { props } = setup();
    await drop([{}, {}], [], props.field);
    expect(props.field.input.onChange).toBeCalledWith({
      errors: [{
        code: 'one file',
        message: 'only one file can be uploaded',
      }],
    });
  });

  test('acceptedFiles.length === 1 calls base64File', async () => {
    // eslint-disable-next-line no-undef
    const readAsDataURLSpy = jest.spyOn(FileReader.prototype, 'readAsDataURL');
    const { props } = setup();
    // eslint-disable-next-line no-undef
    await drop([new Blob()], [], props.field);
    expect(readAsDataURLSpy).toBeCalled();
    expect(props.field.input.onChange).toBeCalled();
  });
});
