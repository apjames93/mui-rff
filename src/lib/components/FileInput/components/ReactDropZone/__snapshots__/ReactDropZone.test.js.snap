// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <input
    accept="image/jpeg, image/png, application/pdf"
    autoComplete="off"
    multiple={true}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
  <h6
    className="MuiTypography-root MuiTypography-subtitle1"
  >
    Drag 'n' drop some files here, or click to select files
  </h6>
</div>
`;
